projects/ion/src/lib/tooltip/tooltip.service.ts

```diff
@@ -0,0 +1,131 @@
+import { Injectable } from '@angular/core';
+import { Subject } from 'rxjs';
+import { TooltipPosition } from '../core/types';
+
+export type RepositionData = {
+  tooltipLeft: number;
+  tooltipHeight: number;
+  hostPosition: Partial<DOMRect>;
+  screenWidth: number;
+  screenHeight: number;
+};
+
+export type tooltipPositionChecks = {
+  [x in TooltipPosition]?: () => boolean;
+};
+
+@Injectable({
+  providedIn: 'root',
+})
+export class TooltipService {
+  reposition = new Subject();
+  hostPosition: Partial<DOMRect>;
+  screenSize: HTMLElement;
+  tootipCoordinates: DOMRect;
+  elementPadding = 16;
+  elementMaxWidth = 208 + this.elementPadding;
+
+  public getNewPosition(): TooltipPosition {
+    const { left, height } = this.tootipCoordinates;
+    const { clientWidth, clientHeight } = this.screenSize;
+
+    if (!this.hostPosition) {
+      return;
+    }
+
+    const repositionData: RepositionData = {
+      tooltipLeft: left,
+      tooltipHeight: height,
+      hostPosition: this.hostPosition,
+      screenWidth: clientWidth,
+      screenHeight: clientHeight,
+    };
+    const positions: tooltipPositionChecks =
+      this.getTooltipPositions(repositionData);
+
+    const newPosition = this.checkPositions(positions);
+
+    return newPosition;
+  }
+
+  public checkPositions(positions: tooltipPositionChecks): TooltipPosition {
+    let newPosition = TooltipPosition.DEFAULT;
+
+    Object.entries(positions).forEach(([position, check]) => {
+      if (check()) {
+        newPosition = position as TooltipPosition;
+      }
+    });
+
+    return newPosition;
+  }
+
+  public getTooltipPositions(
+    respositionData: RepositionData
+  ): tooltipPositionChecks {
+    return {
```

Comment:
To all these functions that are being called at least twice, you can set a const to make this value be calculated only once, and then reuse it on you object creation.